[1.21] Field Registraction Adjustments

Dragon-Seeker opened the pull request:
Adjusts the Field registration API to handle Registry Entries combined with a supplier method of deferred registration

Dragon-Seeker committed:
Adjust FieldRegistrationHandler to allow for MemoizedSupplier's

- Inject supplier into RegistryEntry
- MemoizedSupplier's that caches the first get call made to the supplier

Dragon-Seeker committed:
Rename method within reflection utils

Dragon-Seeker committed:
Adjust AutoRegistryContainer and children implmentations to be

- Made such abstract classes instead of interfaces to allow for direct calls to entry method on AutoRegistryContainer
- Fix MemoizedEntry name
- Move iterateAccessibleStaticFieldsAllowingMemorizedSuppliers to FieldRegistrationHandler

Dragon-Seeker commented:
Note: This will cause breaking changes fyi

BasiqueEvangelist commented on src/main/resources/fabric.mod.json:
```diff
@@ -76,6 +76,9 @@
         "io/wispforest/owo/serialization/OwoComponentTypeBuilder<TT;>"
+      ],
+      "net/minecraft/class_6880": [
+        "java/util/function/Supplier<TT;>"
```
Aren't injected interfaces supposed to have only default methods?

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/registration/reflect/FieldRegistrationHandler.java:
```diff
@@ -88,4 +104,50 @@ private static <T> ReflectionUtils.FieldConsumer<T> createProcessor(ReflectionUt
+                finalValue = (F) fieldValue;
+            }
+
+            if (field.isAnnotationPresent(IterationIgnored.class)) continue;
```
why is this check after the checks for the target field type and for the supplier?

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/registration/reflect/MemoizedEntry.java:
```diff
@@ -0,0 +1,41 @@
+import java.util.function.Supplier;
+
+public class MemoizedEntry<T> implements Supplier<T> {
+
```
is this class needed if there already is a helper method on `AutoRegistryContainer` for making a `MemoizedRegistryEntry`? any people who only want a supplier can still use the registry entry class since you made it extend supplier

BasiqueEvangelist commented on src/main/java/io/wispforest/owo/util/ReflectionUtils.java:
```diff
@@ -2,13 +2,15 @@
 
 import io.wispforest.owo.registration.annotations.AssignedName;
 import io.wispforest.owo.registration.annotations.IterationIgnored;
+import io.wispforest.owo.registration.reflect.MemoizedEntry;
 import org.jetbrains.annotations.ApiStatus;
 import org.jetbrains.annotations.Nullable;
 
 import java.lang.reflect.*;
 import java.util.Locale;
 import java.util.function.Consumer;
 import java.util.function.Function;
+import java.util.function.Supplier;
```
are these imports unused? you didn't add any code that actually uses these classes

Dragon-Seeker committed:
Adjust reflection code to check `IterationIgnored` annotation first before type checks

Dragon-Seeker committed:
Adjust supplier interface injection to be from a defaulted interface variant instead of base supplier

Dragon-Seeker committed:
Adjust path for interface injection

Dragon-Seeker committed:
Add TypedRegistryEntry and add various implementations for common types

- Refactor folder structure
- FIx `register` method on `AutoRegistryContainer` being unavailable due to being package private

Dragon-Seeker committed:
Fix issues with inability to safetly cast type of supplier to MemoizedRegistryEntry

Dragon-Seeker commented:
Closing this as such changes with Item, Block, and EntityType basically make it nearly impossible to use AutoRegistryContainer tbh
